editor.lib2/src/org/netbeans/modules/editor/lib2/highlighting/HighlightingManager.java

```diff
@@ -241,6 +245,7 @@ synchronized HighlightsLayer findLayer(HighlightsContainer container) {
 
         public @Override void propertyChange(PropertyChangeEvent evt) {
             if (evt.getPropertyName() == null || PROP_DOCUMENT.equals(evt.getPropertyName())) {
+                updatePaneFilter();
```

Comment:
What about synchronizaton on the paneFilter member variable ? It seems to be read under document read lock, what if other thread runs rebuildAllLayersImpl ?

Reply:
Good point. After reviewing the existing synchronization approach of the HighlightingManager.Highlighting class, it seems a correct thing to do is to add a "synchronized" keyword to updatePaneFilter(). The other place that accesses paneFilter is rebuildAllLayersImpl, which is only called by rebuildAllLayers, which is also synchronized.

Tested with the manual test code that was attached to the original BugZilla issue.